src/misc/paper-write.ts

```diff
@@ -0,0 +1,224 @@
+  );
+  playersWriting.push(event.player);
+});
+
+/**
+ * Reading paper contents
+ */
+registerEvent(PlayerInteractEvent, async (event) => {
+  if (!isRightClick(event.action)) return;
+  if (event.hand !== EquipmentSlot.HAND) return;
+  if (event.item?.type !== Material.PAPER) return;
+  // Check if player has feather
+  const inventory = event.player.inventory as PlayerInventory;
+  const offHand = inventory.itemInOffHand;
+  if (offHand.type == Material.FEATHER) return;
+  if (playersWriting.includes(event.player)) return;
+  // Check if the paper already has something written on it
+  if (event.item.itemMeta.customModelData != 1) return;
+  if (event.item.itemMeta.lore) {
+    event.player.sendMessage(text('Paperilla lukee:'));
+    event.player.sendMessage(
+      color(ChatColor.WHITE, text(event.item.itemMeta.lore[0])),
+    );
+    return;
+  }
+});
+
+async function handleMessage(msg: ChatMessage) {
+  playersWriting.removeValue(msg.sender);
+  const inventory = msg.sender.inventory as PlayerInventory;
+  const mainHand = inventory.itemInMainHand;
+  const offHand = inventory.itemInOffHand;
+  if (
+    offHand.type !== Material.FEATHER ||
+    mainHand.type !== Material.PAPER ||
+    mainHand?.itemMeta?.lore
+  ) {
+    msg.sender.sendActionBar('§7Kirjoittaminen peruttu.');
+    return;
+  }
+  if (!msg.sender.inventory.contains(Material.INK_SAC, mainHand.amount)) {
+    msg.sender.sendActionBar('§7Sinulla ei ole mustetta tarpeeksi.');
+    return;
+  }
+  const message = msg.content;
+  if (MAX_LENGTH < message.length) {
+    msg.sender.sendMessage(
+      color(ChatColor.GRAY, text('Tämä teksti on liian pitkä. Käytä kirjaa!')),
+    );
+    return;
+  }
+  const itemMeta = mainHand.itemMeta;
+  // Add color to lore
+  itemMeta.lore = [`§7${message}`];
+  itemMeta.customModelData = 1;
+  mainHand.itemMeta = itemMeta;
+  msg.sender.sendActionBar('§7Kirjoittaminen onnistui!');
+  msg.sender.playSound(
+    msg.sender.location,
+    Sound.ENTITY_VILLAGER_WORK_CARTOGRAPHER,
+    1,
+    1,
+  );
+  msg.sender.inventory.removeItem(
+    new ItemStack(Material.INK_SAC, mainHand.amount),
+  );
+}
+
+/**
+ * Alternative interface for detecting writing paper hands.
+ */
+
+export function detectWritingPaper(msg: ChatMessage) {
+  if (playersWriting.includes(msg.sender)) {
+    msg.discard = true;
+    handleMessage(msg);
+  }
+}
+
+GLOBAL_PIPELINE.addHandler('detectWritingPaper', 0, detectWritingPaper);
+
+/**
+ * Put paper inside envelope
+ */
+
+registerEvent(PlayerInteractEvent, async (event) => {
+  if (!isRightClick(event.action)) return;
+  if (event.hand !== EquipmentSlot.HAND) return;
+  if (event.item?.type !== Material.PAPER) return;
+  // Check if player has feather
+  const inventory = event.player.inventory as PlayerInventory;
+  const offHand = inventory.itemInOffHand;
+  if (offHand.type !== Material.PAPER) return;
+  if (offHand.itemMeta.customModelData != 3) return;
+  if (event.item.itemMeta.customModelData != 1) return;
```

Comment:
Could Customitem.event() be used here? Same with other click events with custom paper items